### PR TITLE
[smart pipes] Add support for question-mark tokens 

### PR DESCRIPTION
**Pull request by js-choi**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | No
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | [pending on 7.2.0 docs]
| Any Dependency Changes?  | No
| License                  | MIT

This pull request builds on #8289, which initially implemented parsing of the [stage-0 smart-pipeline proposal](https://github.com/js-choi/proposal-smart-pipelines/). #8289 uses `#` as its “topic token” (aka the “placeholder token”), but this is only a preliminary choice, and the [actual token has not yet been actually decided](https://github.com/tc39/proposal-pipeline-operator/issues/91). This present pull request adds support for switching to the question mark `?` as the topic token using a new `topicToken` configuration option on the `pipelineOperator` parser plugin. Future pull requests will add support for other good possibilities, namely `%` and `@`.

See also #9122 and #9179. CC @mAAdhaTTah, @littledan, @thiagoarrais.

### CODE REVIEW

**Comment by thiagoarrais** [comment]
You might want to update `babel-generator` too. It currently hardcodes `#` as the topic token.

The relevant code is at [types.js/PipelinePrimaryTopicReference](https://github.com/babel/babel/blob/731182eee44a1b47d5232697489a49f09641bccf/packages/babel-generator/src/generators/types.js#L163).

**Comment by js-choi** [comment]
Will do.

**Comment by js-choi** [comment]
On the other hand, it might be worth implementing generator support in a separate  pull request, since, to my understanding, babel-generator is decoupled from babel-parser.

**Comment by js-choi** [comment]
How should the configuration option for generated topic tokens (default `#` but also may be `?`) be designed? There [doesn’t seem to be a plugin-option system in babel-generator](https://www.npmjs.com/package/babel-generator), unlike babel-parser and the transform plugins.

**Comment by nicolo-ribaudo** [comment]
You can add a top-level option, like I did for decorators:https://github.com/babel/babel/blob/c586d4e8cac816941bf2ba559472fd9cc3150921/packages/babel-generator/src/index.js#L55

I understand that this is sub-optimal, but `@babel/generator` has just a few options and is almost never used directly, so a plugin system might be overkill.

File: packages/babel-parser/src/parser/expression.js [hunk 1]
```diff
@@ -2110,6 +2092,82 @@ export default class ExpressionParser extends LValParser {
...
+    );
+    const effectivePluginConfigurationTopicToken =
+      pluginConfigurationTopicToken || defaultTopicToken;
+    return effectivePluginConfigurationTopicToken === topicToken;
+  }
+
+  // Returns the value of the pipelineOperator plugin's topicToken configuration option.
+
+  getPipelineOperatorPluginTopicTokenOption(): ?string {
+    return this.getPluginOption("pipelineOperator", "topicToken");
```
[the remark this answers is not in the document]
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js]: Understood; I’ll inline this.

File: packages/babel-parser/src/parser/expression.js [hunk 2]
```diff
@@ -2110,6 +2092,82 @@ export default class ExpressionParser extends LValParser {
...
+  // topicToken configuration option.
+
+  matchesTopicTokenConfiguration(
+    topicToken: N.PipelineProposedTopicToken,
+  ): boolean {
+    const defaultTopicToken = "#";
+    const pluginConfigurationTopicToken = this.getPluginOption(
+      "pipelineOperator",
+      "topicToken",
+    );
```
[the remark this answers is not in the document]
Comment (js-choi) [reply packages/babel-parser/src/parser/expression.js]: Understood; I’ll replace this.

**Comment by js-choi** [comment]
> You can add a top-level option, like I did for decorators:
> 
> [babel/packages/babel-generator/src/index.js](https://github.com/babel/babel/blob/c586d4e8cac816941bf2ba559472fd9cc3150921/packages/babel-generator/src/index.js#L55)
> Line 55 in [c586d4e](/babel/babel/commit/c586d4e8cac816941bf2ba559472fd9cc3150921)
>  decoratorsBeforeExport: !!opts.decoratorsBeforeExport, 
> I understand that this is sub-optimal, but `@babel/generator` has just a few options and is almost never used directly, so a plugin system might be overkill.

Yes, sounds good. Should I put this functionality in this pull request or in a new pull request?

**Comment by nicolo-ribaudo** [comment]
It's ok in this PR

**Comment by JLHwung** [comment]
Closing this as the smart pipeline proposal is not proposed. As of May 2021, the smart pipeline is replaced by hack pipeline. See also #13191.

**Comment by js-choi** [comment]
Specifically, this pull request has been supplanted by #13191.

**Pull request closed without merging**